### PR TITLE
Remix auth

### PR DESCRIPTION
**Pull request by sverhoeven**
Fixes https://github.com/i-VRESSE/bartender/issues/54

Should be merged together with https://github.com/i-VRESSE/bartender/pull/73

TODO

- [x] store private key for making bartender tokens
- [x] use remix auth for local users with remix-auth-form
- [x] use remix-auth-oauth2 for
  - [x] github
  - [x] orcid sandbox
  - [x] orcid prod
  - [x] egi-checkin
- [x] associate local user with oauth account aka mail matching
- [x] regenerated app/bartender-client against https://github.com/i-VRESSE/bartender/pull/73
- [x] dont generate bartender token every time, but store/fetch from db
- [x] enforce password length with zod
- [x] store users in postgresql instead of sqlite
- [x] Replace expertise level seed with enum as new level needs code change, so why not have levels as code as well.
- [x] document oauth2 provider setup
- [x] docker-compose with bartender and haddock3-webapp dbs + rsa key pair mount
- [x] explain remix + prisma choice 
    - [x] workflow builder works, while in webpack based meta framework the build chokes
    - [x] follow remix blue stack template, for least dev surprises
- [x] Upgraded to latest remix
    - [x] Use v2 route file naming
- [x] use prisma migrations to create tables in db
- [x] Calculate gravatar hash

To test:
1. Start bartender web service from https://github.com/i-VRESSE/bartender/tree/asym-jwt-only branch with public key.
2. Setup webapp according to development instructions in README.md
3. Register + login + give yourself an expertise level
4. Check on build page that topoaa node has options for your selected expertise level
5. Configure social login GitHub, Orcid or EGI and try to login with those accounts. PS. Make sure you a logged out from the webapp before trying to log in again.
6. 
7. Start webapp with docker compose using  Docker instructions in README.m

### CODE REVIEW

**Review by Peter9192 — COMMENTED** [review]
- Profile popup doesn't go away when clicked
- Checkboxes on admin page shouldn't apply directly, but after a save
- Can't change password on first admin user?
- Batch checker doesn't work on admin page (with single user)
-

File: package.json [hunk 1]
```diff
@@ -9,61 +9,80 @@
...
-    "format": "prettier --write .",
+    "format": "prettier --write . '!postgres-data'",
     "generate-client": "openapi-generator-cli generate -g typescript-fetch -i http://localhost:8000/api/openapi.json -o app/bartender-client",
     "lint": "eslint --cache --cache-location ./node_modules/.cache/eslint .",
     "start": "remix-serve build",
     "test": "vitest",
-    "typecheck": "tsc"
+    "typecheck": "tsc",
+    "setup": "prisma generate && prisma migrate deploy && prisma db seed",
+    "docker:dev": "docker compose -f docker-compose.dev.yml up",
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply package.json:19]: Can add ` && docker-compose rm -fsv` to script so after you kill the up command the rm command is run.

**Review by Peter9192 — APPROVED** [review]
Good stuff! I found a few things in the docker compose file, further mostly some clarification suggestions. While browsing to remix docs, I noticed that they mostly like [cookiesessionstorage](https://remix.run/docs/en/main/utils/sessions#createcookiesessionstorage) because you don't need a database for them. So I was wondering, since we have a database anyway, is this really the most suitable choice for storing session info? Furthermore, it would be nice to have a dedicated error page in the same style as the rest of the app. And it would be helpful to clarify (more explicitly) that bartender should not be setting roles, this is done by the web app

File: app/models/user.server.ts [hunk 2]
```diff
@@ -1,170 +1,279 @@
...
+    throw new Error("No expertise levels found");
+  }
+  // cast needed for valibot.enumpType
+  return array as [ExpertiseLevel, ...ExpertiseLevel[]];
 }
 
-export async function listRoles(accessToken: string) {
-  const api = buildRolesApi(accessToken);
-  return await api.listRoles();
+export async function assignExpertiseLevel(
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply app/models/user.server.ts:164]: Nope expertise level is not in token

File: app/components/admin/UserTableRow.tsx [hunk 3]
```diff
@@ -1,47 +1,59 @@
...
+  expertiseLevels,
+  onUpdate,
+  submitting,
+}: IProps) => {
+  const usersExpertiseLevels = user.expertiseLevels;
   return (
     <tr>
+      <td>
+        <input type="checkbox" disabled />
+      </td>
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply app/components/admin/UserTableRow.tsx]: Removed in fcc7c68f3bc6eccaa58526f84f5055f490222ec6

File: app/bartender_token.server.ts [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+export class TokenGenerator {
+  private privateKeyFilename: string;
+  private issuer: string;
+  private lifeSpan: string;
+  private privateKey: KeyLike | undefined = undefined;
+  constructor(
+    privateKeyFilename: string,
+    issuer = "bartender",
+    lifespan = "8h"
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply app/bartender_token.server.ts:21]: The token is never exposed so I don't think it is too long. 
If we had to ban a user then the web app will block any pages that need a token.

File: app/routes/login.tsx [hunk 5]
```diff
@@ -44,26 +65,35 @@ export default function LoginPage() {
...
+          <ErrorMessages path="email" errors={actionData?.errors} />
+        </div>
+        <div>
+          <label>
+            <p>Password</p>
+            <input
+              id="password"
+              name="password"
+              type="password"
+              minLength={8}
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply app/routes/login.tsx]: You are right, removed length check on server and client in 6cfcd2e47f10f8aabb9646f069030e954d4a863b

File: docker-compose.yml [hunk 6]
```diff
@@ -9,10 +9,46 @@ services:
...
         condition: service_started
+      webappdb:
+        condition: service_healthy
     ports:
       - "8080:8080"
+    env_file:
+      - ./.env
     environment:
       BARTENDER_API_URL: "http://bartender:8000"
+      DATABASE_URL: postgresql://postgres:postgres@webappdb:5432/postgres
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply docker-compose.yml:20]: Added in 6cfcd2e47f10f8aabb9646f069030e954d4a863b

File: docs/auth.md [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+- [Authentication \& authorization](#authentication--authorization)
+  - [GitHub login](#github-login)
+  - [Orcid sandbox login](#orcid-sandbox-login)
+  - [Orcid login](#orcid-login)
+  - [EGI Check-in login](#egi-check-in-login)
+
+A user can only submit jobs when he/she is logged in and has at least one expertise level.
+A super user can assign an expertise level to users at http://localhost:3000/admin/users.
+A super user can be made through the admin page (`/admin/users`) or by being the first registered user.
+
```
[the remark this answers is not in the document]
Comment (sverhoeven) [reply docs/auth.md:14]: Disabling super user toggle when you are the only super user would be nice, but confirm is good enough.

**Comment by sverhoeven** [comment]
> Good stuff! I found a few things in the docker compose file, further mostly some clarification suggestions. While browsing to remix docs, I noticed that they mostly like [cookiesessionstorage](https://remix.run/docs/en/main/utils/sessions#createcookiesessionstorage) because you don't need a database for them. So I was wondering, since we have a database anyway, is this really the most suitable choice for storing session info? Furthermore, it would be nice to have a dedicated error page in the same style as the rest of the app. And it would be helpful to clarify (more explicitly) that bartender should not be setting roles, this is done by the web app

The users browser needs to persist something between page loads we are storing the encrypted user id in the cookie just like https://github.com/remix-run/blues-stack/blob/main/app/session.server.ts#L78 . If you would store session id in cookie and session to user mapping as a db table there are more moving parts without much benefit.

For errors you already created #22 so will deal with better error pages when fixing that issue.

Thanks for reviewing this monster pull request and fixing the deployment error I should have caught.

**Pull request merged**